### PR TITLE
fix(stream-caip10-link): better genesis determinism

### PR DESCRIPTION
**Pull request by oed**
Saw in the convo space app that my mainnet account has two caip10-links for the same address. This PR fixes this problem.

image

### CODE REVIEW

**Comment by ukstv** [comment]
Yay!

**Review by ukstv — APPROVED** [review]
🏇

File: packages/stream-caip10-link/src/caip10-link.ts
Comment (stbrody) [inline packages/stream-caip10-link/src/caip10-link.ts:128]: how confident are we that no blockchains will ever have case-sensitive addresses?

It would be a pretty hairy decision to make as a blockchain designer, so here's hoping it never happens...

**Review by stbrody — APPROVED** [review]
LGTM

**Comment by ukstv** [comment]
Added relevant issue on caip repo: https://github.com/pedrouid/caip-js/issues/11

File: packages/stream-caip10-link/src/caip10-link.ts
Comment (oed) [reply packages/stream-caip10-link/src/caip10-link.ts:128]: I'm pretty sure for bitcoin this is not the case: https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md

I do think we need a better strategy mid/long term. Alternatively we add specific logic for each blockchain here, which is not ideal

Comment (stbrody) [reply packages/stream-caip10-link/src/caip10-link.ts:128]: Yeah wait, the CAIP10 spec specifically says that the accountId is case-senstive.  Blindly translating it to lower case feels wrong then and like it's corrupting the links.  If eth addresses are case insensitive, then that needs to be handled in an eth-specific way, not blindly applied to all possible Caip10Links.

**Review by stbrody — CHANGES_REQUESTED** [review]
revoking my lgtm after reviewing the caip10 spec

File: packages/stream-caip10-link/src/caip10-link.ts
Comment (oed) [reply packages/stream-caip10-link/src/caip10-link.ts:128]: Ok, we should do it properly and encode eth addresses as specified by [EIP-55](https://eips.ethereum.org/EIPS/eip-55).

Comment (oed) [reply packages/stream-caip10-link/src/caip10-link.ts:128]: On second thought, doing EIP-55 could potentially cause issues for 3IDs which have already been migrated/created on 3ID Connect. Would like @zachferland to weigh in here if this matters or not.

Comment (stbrody) [reply packages/stream-caip10-link/src/caip10-link.ts]: > On second thought, doing EIP-55 could potentially cause issues for 3IDs which have already been migrated/created on 3ID Connect. Would like @zachferland to weigh in here if this matters or not.

holding off on final review/approval of this PR until this question is addressed.  It does seem like using EIP-55 encoding would be the "better" solution than just lower-casing it (as https://github.com/pedrouid/caip-js/issues/11 is probably the best long-term solution and I imagine that that's what it would do). But if that breaks 3id-connect that's obviously a problem

**Comment by zachferland** [comment]
Looks good, were they not using blockchain-utils-linking library? we already normalize or lowercase addresses for links there, all created through 3id-connect should already be lower case

eip55 would break 3id-connect in current form, and if changed yeah would require some backwards supports to make sure migration paths works fine. In any case it would require creating a new link. So on the flipside, the earlier it is the changed the less people that need to resign a link proof (not huge overhead though)

Would lean towards this fix, since kinda of hard to prioritize changes in 3id-connect for this right now

**Comment by burk3** [comment]
fwiw, we (DNS) have been using blockchain-utils-linking but not 3id-connect and have been creating caip10links and rows in cryptoAccounts w/ EIP-55 addresses. The main, awkward result of this is that when a user with a wallet makes a DID on dns.xyz, attempting to log into (clay.)self.id with the same wallet gives the user the new account flow, but results in the same DID and profile info from DNS still shows up just fine. The user will also end up with two cryptoAccount rows (one with EIP-55 addr, one with all lower) each with their own caip10link.

So yea, extra signature, but no big deal.

**Comment by oed** [comment]
Interesting @burk3 I noticed something similar when I authenticated to another app that was using 3ID Connect. I got the migration flow, but in the end ended up with the same 3ID. If that's the only problem (i.e. no new 3ID is generated) then we should use EIP-55 for sure!
Thoughts @zachferland ?

**Comment by Geo25rey** [comment]
@oed Are you suggesting to only allow EIP-55 addresses in the cryptoAccounts schema? Would that include altering the schema itself to restrict Ethereum based addresses (if that's even possible)?

**Comment by oed** [comment]
@Geo25rey No I don't think that's possible as you say. This is more about a restriction in the caip10-link doctype.

**Comment by Geo25rey** [comment]
@oed does that mean there should be both EIP-55 addresses and lowercase addresses pointing to the same caip10-link stream ID?

**Comment by oed** [comment]
> @oed does that mean there should be both EIP-55 addresses and lowercase addresses pointing to the same caip10-link stream ID?

I think *if* we decide to go with eip55 we should strive to have the entries in the cryptoAccounts also eip55 encoded. There's no way to enforce this on the schema level however.

File: packages/stream-caip10-link/src/caip10-link.ts
Comment (oed) [reply packages/stream-caip10-link/src/caip10-link.ts]: After a conversation with @zachferland we decided that we should go for the EIP-55 path. Will be updating this PR shortly.

**Review by stbrody — APPROVED** [review]
👍

**Comment by stbrody** [comment]
Does this require a corresponding change to IDX to handle the cryptoAccounts definition? @oed @PaulLeCam

**Comment by oed** [comment]
> Does this require a corresponding change to IDX and/or 3id-connect to handle the cryptoAccounts definition?

No.

**Comment by stbrody** [comment]
> No for IDX definition. Might need a change in 3id-connect.

Cool, I was mostly just responding to this earlier comment from you and trying to make sure we didn't forget to follow up if we need to:

> I think if we decide to go with eip55 we should strive to have the entries in the cryptoAccounts also eip55 encoded

**Pull request merged**
